src/py/mat3ra/made/tools/analyze/coordination.py

```diff
@@ -0,0 +1,499 @@
+from enum import Enum
+from typing import Dict, List, Optional
+
+import numpy as np
+from mat3ra.made.material import Material
+from pydantic import BaseModel
+from scipy.spatial._ckdtree import cKDTree
+
+from ..convert import to_pymatgen
+from ..third_party import PymatgenVoronoiNN
+from ..utils import ArrayWithIds, decorator_handle_periodic_boundary_conditions
+from . import BaseMaterialAnalyzer
+
+
+class BondDirectionsTemplatesEnum(List, Enum):
+    OCTAHEDRAL = [[1, 0, 0], [-1, 0, 0], [0, 1, 0], [0, -1, 0], [0, 0, 1], [0, 0, -1]]
+    TETRAHEDRAL = [[1, 1, 1], [-1, -1, 1], [1, -1, -1], [-1, 1, -1]]
+    PLANAR = [[1, 0, 0], [-1, 2, 0], [-1, -2, 0]]
+    LINEAR = [[1, 0, 0], [-1, 0, 0]]
+
+
+class BondDirections(np.ndarray):
+    def __eq__(self, other):
+        if not isinstance(other, BondDirections):
+            return False
+        return self.are_bond_directions_similar(self, other)
+
+    @staticmethod
+    def are_bond_directions_similar(directions1: np.ndarray, directions2: np.ndarray, tolerance: float = 0.1) -> bool:
+        """
+        Check if two bond templates are similar.
+
+        Args:
+            directions1 (np.ndarray): First template of bond vectors.
+            directions2 (np.ndarray): Second template of bond vectors.
+            tolerance (float): Angle tolerance for comparison.
+
+        Returns:
+            bool: True if the templates are similar, False otherwise.
+        """
+        if len(directions1) != len(directions2):
+            return False
+
+        dot_matrix = np.dot(directions1, directions2.T)
+        norms1 = np.linalg.norm(directions1, axis=1)
+        norms2 = np.linalg.norm(directions2, axis=1)
+        cosine_matrix = dot_matrix / np.outer(norms1, norms2)
+        angles_matrix = np.arccos(np.clip(cosine_matrix, -1.0, 1.0))
+
+        unmatched = list(range(len(directions2)))
+        for angle_row in angles_matrix:
+            matches = np.where(angle_row < tolerance)[0]
+            if len(matches) == 0:
+                return False
+            unmatched.remove(matches.tolist()[0])
+
+        return True
+
+
+class CrystalSite(BaseModel):
+    # element: str
+    # coordinate: List[float]
+    nearest_neighbor_vectors: []
+    # coordination_number: int = 0
+    # see https://www.cryst.ehu.es/cgi-bin/cryst/programs/nph-wp-list for an example
+    wyckoff_letter: Optional[str] = None
+
+    @property
+    def coordination_number(self):
+        return len(self.nearest_neighbor_vectors)
+
+
+class CrystalSiteList(ArrayWithIds):
+    values: List[CrystalSite]
+
+
+class MaterialWithCrystalSites(Material):
+    crystal_sites: CrystalSiteList = CrystalSiteList()
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self.nearest_neighbor_vectors = self.get_nearest_neighbors_vectors()
```

Comment:
get_nearest_neighbors_vectors(self, cutoff <from data>)

Reply:
Maybe this? self is already passed
```        
self.nearest_neighbor_vectors = self.get_neighbors_vectors_for_all_sites(cutoff=3.0)
```